segmentstore/server/src/main/java/io/pravega/segmentstore/server/attributes/SegmentAttributeIndex.java

```diff
@@ -0,0 +1,663 @@
+                // Nothing to read.
+                r.complete(new AttributeCollection());
+            } else {
+                AsyncReadResultProcessor.process(
+                        StreamSegmentStorageReader.read(this.attributeSegment.get().handle, lastSnapshotOffset, readLength, this.config.getReadBlockSize(), this.storage),
+                        new AttributeSegmentReader(r, timeout),
+                        this.executor);
+            }
+            return r;
+        }, this.executor);
+        return Futures.exceptionallyCompose(result,
+                ex -> {
+                    ex = Exceptions.unwrap(ex);
+                    if (ex instanceof IOException || ex instanceof StreamSegmentTruncatedException) {
+                        ex = new DataCorruptionException(
+                                String.format("Unable to parse AttributeSegment. LastSnapshot = (Offset=%d, Length=%d), Known Segment Length = %d.",
+                                        getLastSnapshotOffset(), getLastSnapshotLength(), this.attributeSegment.get().getLength()), ex);
+                    }
+                    return Futures.failedFuture(ex);
+                });
+    }
+
+    /**
+     * Creates a new Snapshot, writes it to the AttributeSegment and updates the main Segment's attributes with its location.
+     *
+     * @param newAttributes Any new Attributes to include in this Snapshot that would not already be part of it.
+     * @param mustComplete  Whether the update of the main Segment's attributes must execute successfully in order for this
+     *                      operation to be considered complete. For auto-generated Snapshots, since the Snapshot itself is
+     *                      already written to the AttributeSegment, it is not mandatory for the OperationLog Add &
+     *                      Attribute Segment Truncation operations to complete, as there will be no data loss. However, for
+     *                      Snapshots generated as part of Seals, all of those must complete in order for the operation to
+     *                      be considered successful.
+     * @param timeout       Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will indicate that the operation has completed successfully.
+     */
+    private CompletableFuture<Void> createSnapshot(AttributeCollection newAttributes, boolean mustComplete, Duration timeout) {
+        TimeoutTimer timer = new TimeoutTimer(timeout);
+
+        // The serialization may be invoked multiple times, based on whether the appendConditionally() requires a retry.
+        Supplier<CompletableFuture<ArrayView>> s = () ->
+                readAllSinceLastSnapshot(timer.getRemaining())
+                        .thenApplyAsync(c -> {
+                            c.mergeWith(newAttributes);
+                            return serialize(c);
+                        }, this.executor);
+        return appendConditionally(s, timer)
```

Comment:
Can we actually log (info or debug) the number of attributes we have in a snapshot? It would be good to give the ability to track growth, and I'm not sure metrics is a better path.

Reply:
In line 413, we debug-log the number of bytes written. In line 346, we debug-log the offset & length of the snapshot, in bytes.

I don't pass along the number of attributes to these methods (I'll need new classes) so I can't record those, but if we divide the length of the written data by 24 (UUID+Long), we get a rough estimate of the number of attributes in that write.